### DISCUSSION TITLE
Print multiple Mediaitems on one page

### DISCUSSION THREAD

**Discussion by DominicWindisch** [photobooth-app/photobooth-app#179] [Ideas]
I'm planning on using a rather basic photo printer with a single fixed paper format (10x15 cm). Still, I'd like to print photo stripes (i.e.5x15 cm) or full frame images on it.

Hence the idea to combine two collages to be printed at once to save some paper (2x 5x15 cm = 10x15cm, duh).

A possible solution would be a new API to layout mediaitems using some arrangement (i.e. 2x1 or maybe 2x2, 2x3 etc):

/mediaprocessing/layout/{itemids}/{cols}/{rows}

-> itemids = comma separated IDs
-> cols & rows =  layout matrix

The resulting image could be shown in the UI for confirmation and the printing service would then print it (and possibly trigger deletion afterwards).

What do you think? (Besides the whole other issue of a frontend print select ui 😁)

**Comment by mgineer85** [comment]
I think it could be cool, but also not in focus from my side.
The gallery could have some checkboxes and track selected items similar like photo apps on the smartphone.
That could be sent to an api. Print processpipeline (not yet existing) could take one or more images and process before winding to the printer.

**Reply by DominicWindisch** [reply]
Alright. I'll try to come up with something simple. After all, It's a photo taking app rather than a full fletched photo printer and users should probably not spend a lot of time layouting their prints manually...

**Reply by mgineer85** [reply]
Right. 

Since printing invokes just a command, one could consider to write a simple batch file getting the last X files and print.

**Comment by PixelAlarm** [comment]
Maybe you want to look into this bash Script:
https://gist.github.com/andi34/27a8811a5a1b4dac37b33e348d1866e3

I read it as if its horizontal, it will Print 1 picture in full size.
If its Vertical, it will Print two side by side.

**Reply by DominicWindisch** [reply]
Hey :wave: 

I just started to get into printing. The script isn't quite what I need; it'll duplicate an image instead of printing two different ones. But the core idea is the same - create a temporary collage of the input images which is then forwarded to the printer driver.

imagick appears to have an automatic solution for that: https://stackoverflow.com/a/64433558/4137044 There's got to be an opencv equivalent out there...

**Reply by mgineer85** [reply]
I’d prefer something using python pillow and a simple algorithm to start with like https://stackoverflow.com/questions/74454460/python-pil-image-generate-grid-of-images-of-different-width-height

still not clear to me how multiple files would be selected. Could also be out of scope for a regular PhotoBooth?

**Reply by DominicWindisch** [reply]
What do you think about using [rectpack](https://github.com/secnot/rectpack)?

I'd imagine this workflow:
- when selecting "print" in gallery, user can tick multiple images
- click confirm, selected IDs are transmitted via print API
- in print processing pipeline, the images are grabbed from their id
- as in this [example](https://github.com/secnot/rectpack?tab=readme-ov-file#basic-usage)), image dimensions are added as rectangle to the packer, page size is added as bins
- packer packs the items (takes literally no time for a few images) optimally and fills however many pages required
- using pil, respective images are pasted into a canvas per page
- the pages are fed into the (not yet implemented) print queue

**Reply by mgineer85** [reply]
- rectpack seems unmaintained, so I would rather not add it as dependency 
- printing from gallery (choose multiple images) could combine multiple images, printing from detail image should just print the visible one.
- result of multiple selected images would be one combined image/page to be printed, right? It would be created on the fly and not stored permanently from my understanding.
- I would not implement it in the app myself as I think users could just print multiple Images
- Collages would look awkward in a masonry layout, gifs and video are not (yet) possible to print 
- if you find a reasonable way to implement a full solution, we can add it to the app.

**Reply by DominicWindisch** [reply]
> rectpack seems unmaintained, so I would rather not add it as dependency 

Unfortunately so. I didn't find a maintained alternative that supports rotating the images and automatic distribution amongst multiple canvases (pages).
...

> printing from gallery (choose multiple images) could combine multiple images, printing from detail image should just print the visible one.

Agree.

> result of multiple selected images would be one combined image/page to be printed, right? It would be created on the fly and not stored permanently from my understanding

rectpack could distribute the print items "optimally" amongst however many fixed-size pages necessary. If there are items that can fit on a single page together (e.g. stripes) those would be combined; everything else is printed separately. These intermediate images per page would be created on the fly, yes. 
From a users perspective, I'd expect the selected images to print in full size instead of a "random" collage of miniature versions.

> I would not implement it in the app myself as I think users could just print multiple Images

Yeah, in the end it's a print queue on steroids 😅 printing all the selected items and if some of them happen to fit on a page together, they're printed together.

> Collages would look awkward in a masonry layout, gifs and video are not (yet) possible to print

Yes, these types should be filtered out in print select view. I don't intend to create an "artistically pleasing" layout, but rather a paper saving way of printing multiple items on a fixed page size. Depending on aspect ratio / image size that'll result in some kind of efficient layout which can then be cut into the segments after printing.

> if you find a reasonable way to implement a full solution, we can add it to the app.

I'll figure something out :) I'll try to get rid of the dependency.
